src/communication/channel.hpp

```diff
@@ -57,7 +61,90 @@ class ChannelItf {
     /**
      * Destructor.
      */
-    virtual ~ChannelItf() = default;
+    virtual ~ChannelItf() { }
+};
+
+/**
+ * Channel class.
+ */
+class Channel : public ChannelItf {
+public:
+    /**
+     * Constructor.
+     *
+     * @param transport transport interface.
+     * @param port port number.
+     */
+    Channel(TransportItf* transport, int port);
+
+    /**
+     * Connects to communication channel.
+     *
+     * @return aos::Error.
+     */
+    Error Connect() override;
+
+    /**
+     * Closes current connection.
+     *
+     * @return aos::Error.
+     */
+    Error Close() override;
+
+    /**
+     * Reads data from channel to array.
+     *
+     * @param data buffer where data is placed to.
+     * @param size specifies how many bytes to read.
+     * @return int num read bytes.
+     */
+    int Read(void* buffer, size_t size) override;
+
+    /**
+     * Writes data from array to channel.
+     *
+     * @param data data buffer.
+     * @param size specifies how many bytes to write.
+     * @return int num written bytes.
+     */
+    int Write(const void* buffer, size_t size) override;
+
+    /**
+     * Returns if channel is connected.
+     *
+     * @return bool.
+     */
+    bool IsConnected() const override;
+
+    /**
+     * Waits for read.
+     *
+     * @param buffer buffer where data is placed to.
+     * @param size specifies how many bytes to read.
+     * @return aos::Error.
+     */
+    Error WaitRead(void** buffer, size_t* size);
```

Comment:
Read/Write look like a usual sync functions.  So it's not clear why WaitRead/ReadDone are needed.

Reply:
no, this method is used in channel manager for the reason we wanna reduce extra buffer. So channel manager before passing data to the channel based on the port takes this buffer from the channel fills it and goes to other data.